src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,55 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private T[] elements;
+    private int size;
+    private static final int INITIAL_CAPACITY = 10;
+
+    @SuppressWarnings("unchecked")
+    public MateStack() {
+        elements = (T[]) new Object[INITIAL_CAPACITY];
+        size = 0;
+    }
 
+    public void push(T value) {
+        ensureCapacity();
+        elements[size++] = value;
     }
 
     public T peek() {
-        return null;
+        ensureNotEmpty();
+        return elements[size - 1];
     }
 
     public T pop() {
-        return null;
+        ensureNotEmpty();
+        T value = elements[--size];
+        elements[size] = null;
+        return value;
```

Comment:
Similar to my previous comment, you can create a separate private method `removeValue()` that will handle the logic of removing a value from the stack. Then, in the `pop()` method, you can first call `ensureNotEmpty()` and then `removeValue()`. This will make your code more modular and easier to read and maintain.